### PR TITLE
Probability Distribution and Statistical Functions -- Exponential Distribution Module

### PR DESCRIPTION
**Pull request by Jim-215-Fisher**
This is the second round of probability distribution and statistical functions, continuing from #240. Since the whole modular structure has been changed, each distribution will have its own PR.
This PR contains exponential distributions. Files uploaded are:
src/stdlib_stats_distribtuion_exponential.fypp
src/CMakeLists.txt
src/Makefile.manual
src/tests/test_distribution_exponential.f90
src/tests/CMakeLists.txt
src/tests/Makefile.manual
doc/specs/stdlib_stats_distribution_exponential.md
doc/specs/index.md

### CODE REVIEW

**Comment by Jim-215-Fisher** [comment]
This PR is ready for review.

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
+---
+title: stats_distribution_exponential
+---
+
+# Statistical Distributions -- Exponential Distribution Module
+
+[TOC]
+
+## `rvs_expon` - exponential distribution random variates
```
Comment (jvdp1) [inline doc/specs/stdlib_stats_distribution_exponential.md:9]: I was a bit surprised by the term `expon`. could it be `rvs_exp`? (probably a matter of taste).

File: src/stdlib_stats_distribution_exponential.fypp [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    logical  :: zig_exp_initialized = .false.
+
+    public :: rvs_expon
+    public :: pdf_expon
+    public :: cdf_expon
+
+
+
+    interface rvs_expon
+    !! Version experimental
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:24]:
```suggestion
    !! Version: experimental
```

File: src/stdlib_stats_distribution_exponential.fypp [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+        #:for k1, t1 in RC_KINDS_TYPES
+        module procedure rvs_expon_array_${t1[0]}$${k1}$ !2 dummy variables
+        #:endfor
+    end interface rvs_expon
+
+
+
+    interface pdf_expon
+    !! Version experimental
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:44]:
```suggestion
    !! Version: experimental
```

File: src/stdlib_stats_distribution_exponential.fypp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    !!
+        #:for k1, t1 in RC_KINDS_TYPES
+        module procedure pdf_expon_${t1[0]}$${k1}$
+        #:endfor
+    end interface pdf_expon
+
+
+
+    interface cdf_expon
+    !! Version experimental
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:58]:
```suggestion
    !! Version: experimental
```

File: src/stdlib_stats_distribution_exponential.fypp [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+
+
+contains
+
+    subroutine zigset
+    ! Marsaglia & Tsang generator for random normals & random exponentials.
+    ! Translated from C by Alan Miller (amiller@bigpond.net.au)
+    !
+    ! Marsaglia, G. & Tsang, W.W. (2000) `The ziggurat method for generating
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:79]:
```suggestion
    ! Marsaglia, G. & Tsang, W.W. (2000) 'The ziggurat method for generating
```

File: src/stdlib_stats_distribution_exponential.fypp [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+
+
+
+    #:for k1, t1 in REAL_KINDS_TYPES
+    function rvs_expon_array_${t1[0]}$${k1}$(lambda, array_size) result(res)
+        ${t1}$, intent(in) :: lambda
+        integer, intent(in) :: array_size
+        ${t1}$ :: res(array_size), x, re
+        ${t1}$, parameter :: r = 7.69711747013104972_${k1}$
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:196]: What is this value? How is it computed?

**Review by jvdp1 — COMMENTED** [review]
Thank you @Jim-215-Fisher . Only a few comments. ON overall LGTM.

File: doc/specs/stdlib_stats_distribution_exponential.md (hunk 1 above)
Comment (Jim-215-Fisher) [reply doc/specs/stdlib_stats_distribution_exponential.md:9]: If most people like rvs_exp, I can certainly change it.

File: src/stdlib_stats_distribution_exponential.fypp (hunk 6 above)
Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_exponential.fypp:196]: This parameter value was taken directly from the referenced paper, same as in the normal distribution.

File: doc/specs/stdlib_stats_distribution_exponential.md (hunk 1 above)
Comment (epagone) [reply doc/specs/stdlib_stats_distribution_exponential.md:9]: FWIW, I vote for `rvs_exp`. Among other reasons, I think it's consistent with the Fortran intrinsic exponential function.

Comment (gareth-nx) [reply doc/specs/stdlib_stats_distribution_exponential.md:9]: I agree that `rvs_exp` is a good choice.

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+### Status
+
+Experimental
+
+### Description
+
+An exponentially distributed random variate distribution is the distribution of time between events in a Poisson point process. The inverse scale parameter `lambda` specifies the rate of change.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:17]: Consider replacing with:
* An exponential distribution is the distribution of time between events in a Poisson point process. The inverse scale parameter `lambda` specifies the average time between events, also called the rate of events.

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+Without argument the function returns a standard exponential distributed random variate E(1) with `lambda = 1`.
+
+With single argument, the function returns an exponential distributed random variate E(lambda). For complex arguments, the real and imaginary parts are independent of each other.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:21]: Consider replacing with:
* With a single argument `lambda`, the function returns a random sample from the exponential distribution `E(lambda)`. For complex arguments .... (as before) ...

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+### Status
+
+Experimental
+
+### Description
+
+An exponentially distributed random variate distribution is the distribution of time between events in a Poisson point process. The inverse scale parameter `lambda` specifies the rate of change.
+
+Without argument the function returns a standard exponential distributed random variate E(1) with `lambda = 1`.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:19]: Consider rephrasing as:
* Without argument the function returns a random sample from the standard exponential distribution `E(1)` with `lambda=1`

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+With single argument, the function returns an exponential distributed random variate E(lambda). For complex arguments, the real and imaginary parts are independent of each other.
+
+With two arguments the function returns a rank one array of exponential distributed random variates.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:23]: `exponential distributed` --> `exponentially distributed`

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+Note: the algorithm used for generating normal random variates is fundamentally limited to double precision.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:25]: Should `normal` be `exponential`?

Also, is this a good point to include a reference to the algorithm (or just a statement about what it is)?

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+`result = [[stdlib_stats_distribution_exponential(module):rvs_expon(interface)]]([lambda] [[, array_size]])`
+
+### Class
+
+Function
+
+### Arguments
+
+`lambda`: optional argument has `intent(in)` and is a scalar of type `real` or `complex`.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:37]: I suppose `lambda` has to be nonnegative (and restricted to one quadrant for the complex case). Perhaps worth mentioning that here.

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+Experimental
+
+### Description
+
+The probability density function (pdf) of the single real variable exponential distribution:
+
+$$f(x)=\begin{cases} \lambda e^{-\lambda x} &x\geqslant 0 \\\\ 0 &x< 0\end{}$$
+
+For complex varible (x + y i) with independent real x and imaginary y parts, the joint probability density function is the product of corresponding marginal pdf of real and imaginary pdf (ref. "Probability and Random Processes with Applications to Signal Processing and Communications", 2nd ed., Scott L. Miller and Donald Childers, 2012, p.197):
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:95]: Missing 'a', i.e. `For a complex variable`

**Review by gareth-nx — COMMENTED** [review]
Thanks -- I have a few minor comments on the specs

File: src/stdlib_stats_distribution_exponential.fypp [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+module stdlib_stats_distribution_exponential
+    use stdlib_kinds, only : sp, dp, xdp, qp, int32
+    use stdlib_error, only : error_stop
+    use stdlib_random, only : dist_rand
+    use stdlib_stats_distribution_uniform, only : uni=>rvs_uniform
+
+    implicit none
+    private
+
+    real(dp), parameter  :: ONE = 1.0_dp
```
Comment (gareth-nx) [inline src/stdlib_stats_distribution_exponential.fypp:12]: As this is only used in a single subroutine, I suggest making it a local parameter inside that.

File: src/stdlib_stats_distribution_exponential.fypp [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+    #:for k1, t1 in REAL_KINDS_TYPES
+    function rvs_expon_array_${t1[0]}$${k1}$(lambda, array_size) result(res)
+        ${t1}$, intent(in) :: lambda
+        integer, intent(in) :: array_size
+        ${t1}$ :: res(array_size), x, re
+        ${t1}$, parameter :: r = 7.69711747013104972_${k1}$
+        integer :: jz, iz, i
+
+        if(lambda <= 0.0_${k1}$) call error_stop("Error(rvs_expon_array): Exp" &
+            //"oonential distribution lambda parameter must be greater than zero")
```
Comment (gareth-nx) [inline src/stdlib_stats_distribution_exponential.fypp:200]: One too many `o` in `Exponential`

**Review by gareth-nx — COMMENTED** [review]
Thanks

**Comment by Jim-215-Fisher** [comment]
accidental delete.

**Comment by jvdp1** [comment]
@Jim-215-Fisher For some reasons you also deleted the branch `Distribution-Exponential`, and so, we cannot reopen this PR. 
However, the branch seems to still exist in your repo. So, I suggest to reopen a new PR for this module.

**Comment by awvwgk** [comment]
You can always recover the branch, however we might not be able to reopen this PR. Unless you haven't done so yet use the following the commands to recover the latest state of this branch:

```
git fetch git@github.com:fortran-lang/stdlib ff0da265f5fcf285ce04e15a09625edaefecf51c
git checkout FETCH_HEAD
git switch -c Distribution-Exponential
git push git@github.com:Jim-215-Fisher/stdlib Distribution-Exponential
```

**Comment by jvdp1** [comment]
I checked @Jim-215-Fisher repo and it seems that there is a branch called [`Distribution-eponential`](https://github.com/Jim-215-Fisher/stdlib/tree/Distribution-Exponential), with this last commit: `212bc33dd7b1f43a59a8d904c6a17a1e4aae3eab` called "convert expon name to exp and some words change" which does not seem to have been psuhed in this PR.

Therefore, I suggest to follow @awvwgk procedure. However, I would give it anoter name (e.g., `Distribution-Expornential-1`) and then to merge both branches if they diverge, to be sure you don't loose something.

**Comment by Jim-215-Fisher** [comment]
This branch has been updated. All the corrections have been made.

**Review by gareth-nx — APPROVED** [review]
LGTM, thanks

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+Experimental
+
+### Description
+
+An exponential distribution is the distribution of time between events in a Poisson point process. The inverse scale parameter `lambda` specifies the average time between events, also called the rate of events.
+
+Without argument the function returns a random sample from the standard exponential distribution `E(1)` with `lambda = 1`.
+
+With single argument, the function returns a random sample from the exponential distribution `E(lambda)`. For complex arguments, the real and imaginary parts are independent of each other.
```
Comment (jvdp1) [inline doc/specs/stdlib_stats_distribution_exponential.md:21]:
```suggestion
With a single argument, the function returns a random sample from the exponential distribution `E(lambda)`.
For complex arguments, the real and imaginary parts are sampled independently of each other.
```

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+Experimental
+
+### Description
+
+The probability density function (pdf) of the single real variable exponential distribution:
+
+$$f(x)=\begin{cases} \lambda e^{-\lambda x} &x\geqslant 0 \\\\ 0 &x< 0\end{}$$
+
+For a complex varible (x + y i) with independent real x and imaginary y parts, the joint probability density function is the product of corresponding marginal pdf of real and imaginary pdf (ref. "Probability and Random Processes with Applications to Signal Processing and Communications", 2nd ed., Scott L. Miller and Donald Childers, 2012, p.197):
```
Comment (jvdp1) [inline doc/specs/stdlib_stats_distribution_exponential.md:95]:
```suggestion
For a complex variable (x + y i) with independent real x and imaginary y parts, the joint probability density function
is the product of the corresponding marginal pdf of real and imaginary pdf (for more details, see
"Probability and Random Processes with Applications to Signal Processing and Communications", 2nd ed., Scott L. Miller and Donald Childers, 2012, p.197):
```

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    print *, exp_pdf((1.5,1.0), scale)
+    ! a complex expon probability density function at (1.5,1.0) with real part
+    !of lambda=1.0 and imaginary part of lambda=2.0
+
+! 6.03947677E-02
+
+end program demo_exponential_pdf
+```
+
+## `cdf_exp` - exponential distribution cumulative distribution function
```
Comment (jvdp1) [inline doc/specs/stdlib_stats_distribution_exponential.md:163]:
```suggestion
## `cdf_exp` - exponential cumulative distribution function
```

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+
+Experimental
+
+### Description
+
+Cumulative distribution function (cdf) of the single real variable exponential distribution:
+
+$$F(x)=\begin{cases}1 - e^{-\lambda x} &x\geqslant 0 \\\\ 0 &x< 0\end{}$$
+
+For a complex variable (x + y i) with independent real x and imaginary y parts, the joint cumulative distribution function is the product of corresponding marginal cdf of real and imaginary cdf (ref. "Probability and Random Processes with Applications to Signal Processing and Communications", 2nd ed., Scott L. Miller and Donald Childers, 2012, p.197):
```
Comment (jvdp1) [inline doc/specs/stdlib_stats_distribution_exponential.md:175]:
```suggestion
For a complex variable (x + y i) with independent real x and imaginary y parts, the joint cumulative distribution
function is the product of corresponding marginal cdf of real and imaginary cdf (for more details, see
"Probability and Random Processes with Applications to Signal Processing and Communications", 2nd ed., Scott L. Miller and Donald Childers, 2012, p.197):
```

File: src/stdlib_stats_distribution_exponential.fypp [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+        module procedure pdf_exp_${t1[0]}$${k1}$
+        #:endfor
+    end interface pdf_exp
+
+
+
+    interface cdf_exp
+    !! Version experimental
+    !!
+    !! Exponential Distribution Cumulative Distribution Function
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:59]:
```suggestion
    !! Exponential Cumulative Distribution Function
```

File: src/stdlib_stats_distribution_exponential.fypp [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+        #:endfor
+    end interface cdf_exp
+
+
+
+
+
+contains
+
+    subroutine zigset
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:74]: Could it be `pure`?

File: src/stdlib_stats_distribution_exponential.fypp [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+            we(i) = de / M2
+        end do
+        zig_exp_initialized = .true.
+    end subroutine zigset
+
+
+
+
+    #:for k1, t1 in REAL_KINDS_TYPES
+    function rvs_exp_0_${t1[0]}$${k1}$( ) result(res)
```
Comment (jvdp1) [inline src/stdlib_stats_distribution_exponential.fypp:115]: Could it be `pure`?

**Review by jvdp1 — COMMENTED** [review]
In general LGTM. Only a few minor comments. Thank you for this work.

File: src/stdlib_stats_distribution_exponential.fypp (hunk 21 above)
Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_exponential.fypp:74]: No. The zigset subroutine changes the state of module variable `zig_initialized` and other variables. If we store the table of values accessed by the ziggurat algorithm as module variables, we might eliminate zigset subroutine and make exponential/normal random variables generation functions pure and capable of multithread/multiprocess.

File: src/stdlib_stats_distribution_exponential.fypp (hunk 22 above)
Comment (Jim-215-Fisher) [reply src/stdlib_stats_distribution_exponential.fypp:115]: Same reason as above.

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+
+### Arguments
+
+`lambda`: optional argument has `intent(in)` and is a scalar of type `real` or `complex`. The value of `lambda` has to be non-negative.
+
+`array_size`: optional argument has `intent(in)` and is a scalar of type `integer` with default kind.
+
+### Return value
+
+The result is a scalar or rank one array with a size of `array_size`, and as the same type of `lambda`.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:44]: Suggest to finish with "and has the same type as `lambda`."

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+
+`x`: has `intent(in)` and is a scalar of type `real` or `complex`.
+
+`lambda`: has `intent(in)` and is a scalar of type `real` or `complex`.
+
+All arguments must have the same type.
+
+### Return value
+
+The result is a scalar or an array, with a shape conformable to arguments, and as the same type of input arguments.
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:120]: Suggest to end this with "...arguments, with the same type as the input arguments."

File: doc/specs/stdlib_stats_distribution_exponential.md [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
```
Comment (gareth-nx) [inline doc/specs/stdlib_stats_distribution_exponential.md:202]: Suggest to end this with "...conformable to the arguments, and the same type as the input arguments."

**Review by gareth-nx — APPROVED** [review]
Thanks, I noticed a few very minor details in the documentation, and continue to approve merging this.

**Review by jvdp1 — APPROVED** [review]
Thank you for this PR @Jim-215-Fisher . Currently I have no additional comments.

**Comment by gareth-nx** [comment]
@jvdp1 @epagone @Jim-215-Fisher Does anyone object if we merge this now?

**Comment by jvdp1** [comment]
> @jvdp1 @epagone @Jim-215-Fisher Does anyone object if we merge this now?

Good for me!

**Comment by epagone** [comment]
No objections from me as well!

**Pull request merged**